File: bindings/dart/lib/src/rtsp.dart
// SPDX-License-Identifier: PMPL-1.0-or-later
// Copyright (c) 2026 Jonathan D.A. Jewell (hyperpolymath) <j.d.a.jewell@open.ac.uk>
//
// RTSP protocol types for proven-servers.

/// Method matching the Idris2 ABI tags.
enum Method {
  describe(0),
  setup(1),
  play(2),
  pause(3),
  teardown(4),
  getParameter(5),
  setParameter(6),
  options(7),
  announce(8),
  record(9),
  redirect(10);

  const Method(this.tag);
  final int tag;

  static Method? fromTag(int tag) {
    for (final v in values) {
      if (v.tag == tag) return v;
    }
    return null;
  }
}

/// TransportProtocol matching the Idris2 ABI tags.
enum TransportProtocol {
  rtpAvpUdp(0),
  rtpAvpTcp(1),
  rtpAvpUdpMulticast(2);

  const TransportProtocol(this.tag);
  final int tag;

  static TransportProtocol? fromTag(int tag) {
    for (final v in values) {
      if (v.tag == tag) return v;
    }
    return null;
  }
}

/// SessionState matching the Idris2 ABI tags.
enum SessionState {
  init(0),
  ready(1),
  playing(2),
  recording(3);

  const SessionState(this.tag);
  final int tag;

  static SessionState? fromTag(int tag) {
    for (final v in values) {
      if (v.tag == tag) return v;
    }
    return null;
  }
}

/// StatusCode matching the Idris2 ABI tags.
enum StatusCode {
  statusCode_Ok(0),
  movedPermanently(1),
  movedTemporarily(2),
  badRequest(3),
  unauthorized(4),
  notFound(5),
  statusCode_MethodNotAllowed(6),
  notAcceptable(7),
  sessionNotFound(8),
  internalServerError(9),
  notImplemented(10),
  serviceUnavailable(11);

  const StatusCode(this.tag);
  final int tag;

  static StatusCode? fromTag(int tag) {
    for (final v in values) {
      if (v.tag == tag) return v;
    }
    return null;
  }
}

/// RtspError matching the Idris2 ABI tags.
enum RtspError {
  rtspError_Ok(0),
  invalidSlot(1),
  notActive(2),
  invalidTransition(3),
  rtspError_MethodNotAllowed(4),
  transportError(5),
  sessionExpired(6);

  const RtspError(this.tag);
  final int tag;

  static RtspError? fromTag(int tag) {
    for (final v in values) {
      if (v.tag == tag) return v;
    }
    return null;
  }
}

File: bindings/javascript/src/rtsp.js
// SPDX-License-Identifier: PMPL-1.0-or-later
// Copyright (c) 2026 Jonathan D.A. Jewell (hyperpolymath) <j.d.a.jewell@open.ac.uk>
//
// RTSP protocol types for proven-servers.

/** Method matching the Idris2 ABI tags. */
export const Method = Object.freeze({
  DESCRIBE: 0,
  SETUP: 1,
  PLAY: 2,
  PAUSE: 3,
  TEARDOWN: 4,
  GET_PARAMETER: 5,
  SET_PARAMETER: 6,
  OPTIONS: 7,
  ANNOUNCE: 8,
  RECORD: 9,
  REDIRECT: 10,
});

/** TransportProtocol matching the Idris2 ABI tags. */
export const TransportProtocol = Object.freeze({
  RTP_AVP_UDP: 0,
  RTP_AVP_TCP: 1,
  RTP_AVP_UDP_MULTICAST: 2,
});

/** SessionState matching the Idris2 ABI tags. */
export const SessionState = Object.freeze({
  INIT: 0,
  READY: 1,
  PLAYING: 2,
  RECORDING: 3,
});

/** StatusCode matching the Idris2 ABI tags. */
export const StatusCode = Object.freeze({
  STATUS_CODE_OK: 0,
  MOVED_PERMANENTLY: 1,
  MOVED_TEMPORARILY: 2,
  BAD_REQUEST: 3,
  UNAUTHORIZED: 4,
  NOT_FOUND: 5,
  STATUS_CODE_METHOD_NOT_ALLOWED: 6,
  NOT_ACCEPTABLE: 7,
  SESSION_NOT_FOUND: 8,
  INTERNAL_SERVER_ERROR: 9,
  NOT_IMPLEMENTED: 10,
  SERVICE_UNAVAILABLE: 11,
});

/** RtspError matching the Idris2 ABI tags. */
export const RtspError = Object.freeze({
  RTSP_ERROR_OK: 0,
  INVALID_SLOT: 1,
  NOT_ACTIVE: 2,
  INVALID_TRANSITION: 3,
  RTSP_ERROR_METHOD_NOT_ALLOWED: 4,
  TRANSPORT_ERROR: 5,
  SESSION_EXPIRED: 6,
});
